lib/ASRouterPreferences.jsm

```diff
@@ -0,0 +1,104 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+"use strict";
+
+ChromeUtils.import("resource://gre/modules/Services.jsm");
+
+const PROVIDER_PREF = "browser.newtabpage.activity-stream.asrouter.messageProviders";
+const DEVTOOLS_PREF = "browser.newtabpage.activity-stream.asrouter.devtoolsEnabled";
+
+const DEFAULT_STATE = {
+  _initialized: false,
+  _providers: null,
+  _providerPref: PROVIDER_PREF,
+  _devtoolsEnabled: null,
+  _devtoolsPref: DEVTOOLS_PREF
+};
+
+class _ASRouterPreferences {
+  constructor() {
+    Object.assign(this, DEFAULT_STATE);
+    this._callbacks = new Set();
+  }
+
+  get providers() {
+    if (!this._initialized || this._providers === null) {
```

Comment:
since `ASRouterPreferences` is a singleton, it could be accessed before the observers are added; we shouldn't start caching until observers are added, since the cached value could be out of date